src/type.js

```diff
@@ -0,0 +1,249 @@
+'use strict';
+
+import { core } from 'metal';
+
+const validators = {
+	any: () => true,
+	array: validateType('array'),
+	bool: validateType('boolean'),
+	func: validateType('function'),
+	number: validateType('number'),
+	object: validateType('object'),
+	string: validateType('string'),
+
+	arrayOf: (validator) => {
+		return composeValidator(
+			arrayOfValidatorFactory,
+			validator
+		);
+	},
+	instanceOf: (expectedClass) => {
+		return composeValidator(
+			instanceOfValidatorFactory,
+			expectedClass
+		);
+	},
+	objectOf: (validator) => {
+		return composeValidator(
+			objectOfValidatorFactory,
+			validator
+		);
+	},
+	oneOfType: (arrayOfTypeValidators) => {
+		return composeValidator(
+			oneOfTypeValidatorFactory,
+			arrayOfTypeValidators
+		);
+	},
+	shape: (shape) => {
+		return composeValidator(
+			shapeValidatorFactory,
+			shape
+		);
+	}
+};
+
+/**
+ * Creates a validator that checks the values of an array against a type.
+ * @param {!function} validator Type validator to check each index against.
+ * @param {boolean} noWarn Determines whether warnings should be emitted.
+ * @return {function} Validator.
+ */
+function arrayOfValidatorFactory(validator, noWarn) {
+	return (value, name, context) => {
+		if (!Array.isArray(value)) {
+			return warning('Expected an array.', name, context, noWarn);
+		} else {
+			const testArray = value.every(
+				item => {
+					return validator.noWarn(item, name);
```

Comment:
As far as the `_noWarn` flag, I'm not a huge fan of that. It doesn't seem referentially transparent since you could call that validator with the same args and get a different side effect for two different cases. But I think if we do move the warning to `metal-state`, this also solves the issue so that we don't have to use `composeValidator` or the `_noWarn` flag.

I like the idea of metal-state emitting the error. Makes sense for other users who may want custom warnings. Do you think it makes more sense to check for an instance of `Error()` rather than just a string? This locks it in so that users wouldnt just check for a truthy value `(val) => val`. This way returning an `Error()` is a very intentional use case.